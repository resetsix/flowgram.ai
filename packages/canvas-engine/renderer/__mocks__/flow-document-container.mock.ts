/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { EntityManager } from '@flowgram.ai/core'
import {
  FlowDocument,
  FlowDocumentContainerModule,
  FlowDocumentContribution,
  FlowNodeTransformData,
  FlowNodeTransitionData,
} from '@flowgram.ai/document'
import { Container, decorate, injectable, type interfaces } from 'inversify'

export class FlowDocumentMockRegister implements FlowDocumentContribution {
  registerDocument(document: FlowDocument) {
    document.registerNodeDatas(FlowNodeTransformData, FlowNodeTransitionData)
  }
}

decorate(injectable(), FlowDocumentMockRegister)

export function createDocumentContainer(): interfaces.Container {
  const container = new Container()
  container.load(FlowDocumentContainerModule)
  container.bind(EntityManager).toSelf()
  container.bind(FlowDocumentContribution).to(FlowDocumentMockRegister)
  return container
}

export function createDocument(): FlowDocument {
  return createDocumentContainer().get(FlowDocument)
}
