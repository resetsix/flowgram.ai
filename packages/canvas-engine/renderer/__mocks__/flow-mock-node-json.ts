/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

export const flowJson = {
  nodes: [
    {
      type: 'start',
      id: 'start',
    },
    {
      type: 'mock',
      id: 'mock',
    },
  ],
}
