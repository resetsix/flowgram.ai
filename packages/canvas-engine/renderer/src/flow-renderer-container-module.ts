/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import { ContainerModule } from 'inversify';

import { FlowRendererResizeObserver } from './flow-renderer-resize-observer';
import { FlowRendererRegistry } from './flow-renderer-registry';

export const FlowRendererContainerModule = new ContainerModule(bind => {
  bind(FlowRendererRegistry).toSelf().inSingletonScope();
  bind(FlowRendererResizeObserver).toSelf().inSingletonScope();
});
