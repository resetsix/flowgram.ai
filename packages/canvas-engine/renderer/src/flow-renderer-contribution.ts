/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import type { FlowRendererRegistry } from './flow-renderer-registry';

export const FlowRendererContribution = Symbol('FlowRendererContribution');

export interface FlowRendererContribution {
  registerRenderer?(registry: FlowRendererRegistry): void;
}
