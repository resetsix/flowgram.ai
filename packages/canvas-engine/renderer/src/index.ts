/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

export * from './entities';
export * from './layers';
export * from './flow-renderer-contribution';
export * from './flow-renderer-registry';
export * from './flow-renderer-container-module';

export { ScrollBarEvents } from './utils';
export { MARK_ARROW_ID } from './components/MarkerArrow';
export { MARK_ACTIVATED_ARROW_ID } from './components/MarkerActivatedArrow';
export { useBaseColor } from './hooks/use-base-color';
export { createLines } from './components/LinesRenderer';
