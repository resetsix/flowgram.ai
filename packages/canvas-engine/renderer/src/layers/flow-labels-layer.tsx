/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import React from 'react';

import { throttle } from 'lodash-es';
import { inject, injectable } from 'inversify';
import { domUtils } from '@flowgram.ai/utils';
import {
  FlowDocument,
  FlowDocumentTransformerEntity,
  FlowNodeEntity,
  FlowNodeTransitionData,
  FlowRendererStateEntity,
} from '@flowgram.ai/document';
import { Layer, observeEntity, observeEntityDatas } from '@flowgram.ai/core';

import { useBaseColor } from '../hooks/use-base-color';
import { FlowRendererRegistry } from '../flow-renderer-registry';
import { createLabels } from '../components/LabelsRenderer';

@injectable()
export class FlowLabelsLayer extends Layer {
  @inject(FlowDocument) readonly document: FlowDocument;

  @inject(FlowRendererRegistry) readonly rendererRegistry: FlowRendererRegistry;

  node = domUtils.createDivWithClass('gedit-flow-labels-layer');

  @observeEntity(FlowDocumentTransformerEntity)
  readonly documentTransformer: FlowDocumentTransformerEntity;

  @observeEntity(FlowRendererStateEntity)
  readonly flowRenderState: FlowRendererStateEntity;

  /**
   * 监听 transition 变化
   */
  @observeEntityDatas(FlowNodeEntity, FlowNodeTransitionData)
  _transitions: FlowNodeTransitionData[];

  get transitions(): FlowNodeTransitionData[] {
    return this.document.getRenderDatas<FlowNodeTransitionData>(FlowNodeTransitionData);
  }

  /**
   * 监听缩放，目前采用整体缩放
   * @param scale
   */
  onZoom(scale: number) {
    this.node!.style.transform = `scale(${scale})`;
  }

  /**
   * 可视区域变化
   */
  onViewportChange: ReturnType<typeof throttle> = throttle(() => {
    this.render();
  }, 100);

  onReady() {
    // 图层顺序调整：节点 > label > 线条
    // 节点 z-index: 10
    this.node.style.zIndex = '9';
  }

  /**
   * 监听readonly和 disabled 状态 并刷新layer, 并刷新
   */
  onReadonlyOrDisabledChange() {
    this.render();
  }

  render() {
    const labels: JSX.Element[] = [];
    if (this.documentTransformer?.loading) return <></>;
    this.documentTransformer?.refresh?.();
    const { baseActivatedColor, baseColor } = useBaseColor();
    const isViewportVisible = this.config.isViewportVisible.bind(this.config);
    this.transitions.forEach((transition) => {
      createLabels({
        data: transition,
        rendererRegistry: this.rendererRegistry,
        isViewportVisible,
        labelsSave: labels,
        getLabelColor: (activated) => (activated ? baseActivatedColor : baseColor),
      });
    });
    // 这里采用扁平化的 react 结构性能更高
    return <>{labels}</>;
  }
}
