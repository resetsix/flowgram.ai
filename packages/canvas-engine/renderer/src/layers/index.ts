/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

export * from './flow-nodes-transform-layer';
export * from './flow-nodes-content-layer';
export * from './flow-lines-layer';
export * from './flow-labels-layer';
export * from './flow-debug-layer';
export * from './flow-scroll-bar-layer';
export * from './flow-drag-layer';
export * from './flow-selector-box-layer';
export * from './flow-selector-bounds-layer';
export * from './flow-context-menu-layer';
export * from './flow-scroll-limit-layer';
