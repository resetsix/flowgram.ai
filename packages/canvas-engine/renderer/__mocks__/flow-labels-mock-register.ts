/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import {
  type FlowNodeRegistry,
  FlowTransitionLabelEnum,
} from '@flowgram.ai/document';
import { FlowTextKey } from '../src/flow-renderer-registry';

/**
 * 动态接入 mock register，测试 labels
 */
export const FlowLabelsMockRegister: FlowNodeRegistry = {
  type: 'mock',
  getLabels(transition) {
    return [
      {
        type: FlowTransitionLabelEnum.BRANCH_DRAGGING_LABEL,
        offset: transition.transform.outputPoint,
        props: {
          side: 'left',
        }
      },
      {
        type: FlowTransitionLabelEnum.BRANCH_DRAGGING_LABEL,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.COLLAPSE_LABEL,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.COLLAPSE_ADDER_LABEL,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.COLLAPSE_ADDER_LABEL,
        offset: transition.transform.outputPoint,
        props: {
          activateNode: {
            getData: () => ({ hovered: true }),
            isVertical: true,
          }
        }
      },
      {
        type: FlowTransitionLabelEnum.COLLAPSE_ADDER_LABEL,
        offset: transition.transform.outputPoint,
        props: {
          activateNode: {
            getData: () => ({ hovered: true }),
            isVertical: true,
          },
        }
      },
      {
        type: FlowTransitionLabelEnum.TEXT_LABEL,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.TEXT_LABEL,
        renderKey: FlowTextKey.LOOP_WHILE_TEXT,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.TEXT_LABEL,
        renderKey: FlowTextKey.CATCH_TEXT,
        props: {
          style: {
            width: 100
          }
        },
        rotate: '90deg',
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.CUSTOM_LABEL,
        offset: transition.transform.outputPoint,
      },
      {
        type: FlowTransitionLabelEnum.CUSTOM_LABEL,
        renderKey: FlowTextKey.LOOP_WHILE_TEXT,
        offset: transition.transform.outputPoint,
      },
      {
        type: 'unknown' as any,
        offset: transition.transform.outputPoint,
      },
    ];
  },
};
