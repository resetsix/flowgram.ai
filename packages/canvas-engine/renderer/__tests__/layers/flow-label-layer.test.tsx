/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

import React from 'react';

import { beforeEach, describe, expect, it, vi } from 'vitest';
import { Container, decorate, injectable, type interfaces } from 'inversify';
import { render } from '@testing-library/react';
import {
  FlowDocument,
  FlowDocumentContainerModule,
  FlowDocumentContribution,
} from '@flowgram.ai/document';
import {
  createDefaultPlaygroundConfig,
  PlaygroundConfig,
  PlaygroundContainerModule,
} from '@flowgram.ai/core';

import { FlowLabelsLayer } from '../../src/layers/flow-labels-layer';
import { FlowRendererRegistry, FlowTextKey } from '../../src/flow-renderer-registry';
import { FlowRendererContribution } from '../../src/flow-renderer-contribution';
import { FlowRendererContainerModule } from '../../src/flow-renderer-container-module';
import { flowJson } from '../../__mocks__/flow-mock-node-json';
import { FlowLabelsMockRegister } from '../../__mocks__/flow-labels-mock-register';
import { FlowDocumentMockRegister } from '../../__mocks__/flow-document-container.mock';

class FlowRenderMockRegister implements FlowRendererContribution {
  registerRenderer(registry: FlowRendererRegistry): void {
    registry.registerLayers(FlowLabelsLayer);
  }
}

decorate(injectable(), FlowRenderMockRegister);

function createDocumentContainer(): interfaces.Container {
  const container = new Container();
  container.load(FlowDocumentContainerModule);
  container.bind(FlowDocumentContribution).to(FlowDocumentMockRegister);
  return container;
}

// layer 层 drag entity 单测
describe('flow-label-layer', () => {
  let container = createDocumentContainer();
  let document: FlowDocument;
  let registry: FlowRendererRegistry;

  beforeEach(() => {
    container = createDocumentContainer();
    container.load(FlowRendererContainerModule);
    container.load(PlaygroundContainerModule);
    container.bind(FlowRendererContribution).to(FlowRenderMockRegister);
    container.bind(PlaygroundConfig).toConstantValue(createDefaultPlaygroundConfig());

    document = container.get<FlowDocument>(FlowDocument);
    document.init();
    document.registerFlowNodes(
      FlowLabelsMockRegister, // 通过 getLabel 方法 mock label
    );
    document.fromJSON(flowJson);
    registry = container.get<FlowRendererRegistry>(FlowRendererRegistry);
    registry.init();
  });

  // 测试初始化
  it('test ready', () => {
    registry.pipeline.renderer.layers.forEach(layer => {
      layer?.onReady?.();
      expect(layer.node.style.zIndex).toEqual('9');
    });
  });

  // 缩放
  it('test zoom', () => {
    registry.pipeline.renderer.layers.forEach(layer => {
      layer?.onZoom?.(2);
      expect(layer.node!.style.transform).toEqual('scale(2)');
    });
  });

  // FIXME: render 单测目前不全
  // 渲染
  it('test render', () => {
    vi.mock('@flowgram.ai/core', async importOriginal => {
      const contextMaker = {
        makeFormItemMaterialContext: vi.fn().mockReturnValue('mock-context'),
        isDragBranch: true,
        labelSide: 'left',
        isDroppableBranch: () => true,
        dropNodeId: 'mock',
        dragging: true,
        isDroppableNode: () => true,
      };
      return {
        // @ts-ignore
        ...(await importOriginal()),
        // mock Adder 组件里的 useService
        useService: vi.fn().mockReturnValue(contextMaker),
      };
    });
    const res: (JSX.Element | undefined)[] = [];

    registry.pipeline.renderer.layers.forEach(layer => {
      const render = (layer as any)?._render?.bind(layer);
      // mock rendererRegistry
      // @ts-ignore
      layer.rendererRegistry = {
        getRendererComponent: () => ({
          renderer: () => null,
        }),
        getText: key => {
          if (key === FlowTextKey.LOOP_WHILE_TEXT) {
            return '123';
          }
          return;
        },
      };
      res.push(render());
    });

    const app = render(<>{res}</>);
    expect(app.asFragment()).toMatchSnapshot();
  });
});
