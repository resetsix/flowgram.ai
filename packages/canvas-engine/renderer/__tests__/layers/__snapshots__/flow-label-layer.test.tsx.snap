// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flow-label-layer > test render 1`] = `
<DocumentFragment>
  <div
    data-label-id="start"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-adder"
      data-from="start"
      data-testid="sdk.flowcanvas.line.adder"
      data-to="mock"
      style="width: 280px; height: 0px; display: flex; justify-content: center; align-items: center;"
    />
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-branch-draggable-adder"
    />
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  />
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-collapse"
      style="width: 280px; height: 32px; display: flex; justify-content: center; align-items: center;"
    />
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-collapse-adder"
      style="display: flex;"
    >
      <div
        class="flow-canvas-adder"
        data-from="mock"
        data-testid="sdk.flowcanvas.line.adder"
        data-to=""
        style="width: 40px; height: 32px; display: flex; justify-content: center; align-items: center;"
      />
    </div>
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-collapse-adder"
    >
      <div
        class="flow-canvas-collapse"
        style="width: 280px; height: 20px; display: flex; justify-content: center; align-items: flex-end;"
      />
      <div
        class="flow-canvas-adder"
        data-from="mock"
        data-testid="sdk.flowcanvas.line.adder"
        data-to=""
        style="width: 280px; height: 20px; display: flex; justify-content: center; align-items: center;"
      />
    </div>
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      class="flow-canvas-collapse-adder"
    >
      <div
        class="flow-canvas-collapse"
        style="width: 280px; height: 20px; display: flex; justify-content: center; align-items: flex-end;"
      />
      <div
        class="flow-canvas-adder"
        data-from="mock"
        data-testid="sdk.flowcanvas.line.adder"
        data-to=""
        style="width: 280px; height: 20px; display: flex; justify-content: center; align-items: center;"
      />
    </div>
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      data-label-id="mock"
      style="font-size: 12px; color: rgb(187, 191, 196); text-align: center; white-space: nowrap; background-color: var(--g-editor-background); line-height: 20px;"
    >
      123
    </div>
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  >
    <div
      data-label-id="mock"
      style="font-size: 12px; color: rgb(187, 191, 196); text-align: center; white-space: nowrap; background-color: var(--g-editor-background); line-height: 20px; width: 100px; transform: rotate(90deg);"
    >
      catch-text
    </div>
  </div>
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  />
  <div
    data-label-id="mock"
    style="position: absolute; left: 0px; top: 0px; transform: translate(-50%, -50%);"
  />
</DocumentFragment>
`;
